specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+Usage of circuit relay, AutoNAT or AutoRelay will be specifically re-examined closer to the time.
+
+## Encryption and identification
+
+#### Interop
+
+[SecIO](https://github.com/libp2p/specs/tree/master/secio) with `secp256k1` identities will be used for initial interoperability testing.
+
+The following SecIO parameters MUST be supported by all stacks:
+
+-  Key agreement: ECDH-P256.
+-  Cipher: AES-128.
+-  Digest: SHA256.
+
+#### Mainnet
+
+[Noise Framework](http://www.noiseprotocol.org/) handshakes will be used for mainnet. libp2p Noise support [is in the process of being standardised](https://github.com/libp2p/specs/issues/195) in the libp2p project.
+
+Noise support will presumably include IX, IK and XX handshake patterns, and may rely on Curve25519 keys, ChaCha20 and Poly1305 ciphers, and SHA256 as a hash function. These aspects are being actively debated in the referenced issue [ETH 2.0 implementers are welcome to comment and contribute to the discussion.]
+
+## Protocol Negotiation
+
+#### Interop
+
+Connection-level and stream-level (see the rationale section below for explanations) protocol negotiation MUST be conducted using [multistream-select v1.0](https://github.com/multiformats/multistream-select/). Its protocol ID is: `/multistream/1.0.0`.
+
+#### Mainnet
+
+Clients MUST support [multistream-select 1.0](https://github.com/multiformats/multistream-select/) and MAY support [multiselect 2.0](https://github.com/libp2p/specs/pull/95). Depending on the number of clients that have implementations for multiselect 2.0 by mainnet, [multistream-select 1.0](https://github.com/multiformats/multistream-select/) may be phased out.
+
+## Multiplexing
+
+During connection bootstrapping, libp2p dynamically negotiates a mutually supported multiplexing method to conduct parallel conversations. This applies to transports that are natively incapable of multiplexing (e.g. TCP, WebSockets, WebRTC), and is omitted for capable transports (e.g. QUIC).
+
+Two multiplexers are commonplace in libp2p implementations: [mplex](https://github.com/libp2p/specs/tree/master/mplex) and [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). Their protocol IDs are, respectively: `/mplex/6.7.0` and `/yamux/1.0.0`.
+
+Clients MUST support [mplex](https://github.com/libp2p/specs/tree/master/mplex) and MAY support [yamux](https://github.com/hashicorp/yamux/blob/master/spec.md). If both are supported by the client, yamux must take precedence during negotiation. See the Rationale section of this document for tradeoffs.
+
+# ETH2 network interaction domains
+
+## Constants
+
+This section outlines constants that are used in this spec.
+
+- `RQRP_MAX_SIZE`: The max size of uncompressed req/resp messages that clients will allow.
```

Comment:
nit: what is RQRP? Can this be MSG_MAX_SIZE?

Reply:
looks like an abbreviation for ReQuestResPonse

Reply:
Guilty as charged.`MSG` is ambiguous though; other layers also deal with messages.

Reply:
I'm not too happy with Req/Resp in general, it's quite the mouthful, happy to hear alternatives.. `DM_MAX_SIZE` is one for direct messaging between peers, as opposed to gossip/broadcast.

Reply:
Naming things is hard 😛 How about we think outside the box of messaging patterns, and describe this domain functionally:

1. "the direct domain" / `DIRECT_MAX_SIZE`
2. "the service domain" / `SVC_MAX_SIZE`
3. "the operation domain" / `OP_MAX_SIZE`

Personally, I like _the service domain_, as it accurately describes what functionality this offers: peers offer services to one another, where one of those services is the synchronisation facility.

WDYT?

Reply:
Did another round of thinking here, and I'm leaning towards `REQ_MAX_SIZE` for the req/resp domain - services sounds like web services or the like which are quite different in style from this - in particular web services tend to be unidirectional.

We have the pub/sub duality and I suspect the req/resp is another example of the same, leaning on the terminology of p2p message buses like 0mq.

Reply:
`REQ_` => my understanding was that this limitation applied to responses too.

Reply:
yeah, but it's the request domain which implies a response. not perfect, agree.

Reply:
so whats wrong w/ `REQ_RESP_MAX_SIZE`? I think the stumbling block is less a perfect name and more that `RQRP` is not a commonly used abbreviation (as far as i have seen)

Reply:
> in particular web services tend to be unidirectional

That isn't true. Both REST and SOAP are inherently bidirectional; with or without payloads, they're synchronous.

**Service domain** describes _what_ this domain does, **Req/resp domain** describes _how_ it does it in terms of messaging patterns.

We name other domains with the _what_, not with the _how_, so applying that consistency here would be fair.

Anyway, I'm fine whichever way; I do confess I never really liked req/resp to be honest 😛

Reply:
at this point, I guess it's less about choosing a good name and more about choosing one that is not too bad :) `REQ_RESP_MAX_SIZE` gets bonus points for staying true to eth2's penchant for excessively long names.

I hold commit rights and take bribes.